pettingzoo/classic/tictactoe/board.py

```diff
@@ -1,79 +1,113 @@
+class BadTicTacToeMoveException(Exception):
```

Comment:
This is a sloppy way of doing it, you should do it the same way the other envs handle illegal moves and action masks (terminate illegal wrapper)

Reply:
Illegal moves are still handled via the wrapper and masks if using the wrapper (still the default). This only gets triggered if using the raw env. The previous version of the raw env failed with an assertion on an illegal move. This one fails with an exception. I think that's a better approach.

Reply:
Is the problem the fact that it's an exception or that it is a custom exception? If the latter, it can be replaced with a ValueError and still accomplish the same thing.

If the former, I was of the understanding that using assert to verify proper input is a misuse. Also, assertions can be disabled by -O which will change the program's behavior. That's one of the reasons the [google style guide](https://google.github.io/styleguide/pyguide.html#244-decision) (2nd bullet) says not to use them in this way.

Custom exceptions have value (though the tictactoe one may be too specific). If there was a consistently used IllegalMoveException across envs, I think the terminate illegal wrapper could be implemented clearer, more robustly, and with lower overhead (just catch that exception). You can't reasonably do that with an assert.

Reply:
That’s fair, I think in that case we should try to make all of the envs do exceptions in this same way, it’s just skippy to have one env function differently. One thing I’ve seen used for these kinds of input type checking and stuff is pydantic (used by langchain for example) which makes the program fail with an error if the type is wrong, don’t think it would apply for illegal moves but a lot of the assertions we have are about the types of things, it’s arguable that we should do that and have more strict typing, but that worthless be a lot of work and diverge from other farama projects. Have been thinking about it for a while though and if we’re going to updating an env anyways and planning to do another breaking change I may just pull the trigger. Will probably talk to other devs to see their thoughts.

Reply:
Ok. Rather than bog this down, I'll change the exceptions back to asserts. I'll also add an issue when I get a chance proposing changing all the asserts for illegal moves to a (new) IllegalMoveError and changing the handling of the illegal move wrapper to match it. People can give feedback there.

Reply:
OK sounds good. I may make an issue proposing pydantic and using TNT logging package, or other more modern ways of handling things vs asserts

Reply:
Made a post in the internal core devs channel about it will see what ppl say